workflow/lib/persistent_dict.py

```diff
@@ -0,0 +1,891 @@
+"""Generic persistent, concurrent dictionary-like facility."""
+
+"""
+Copied into nextstrain/ncov workflow from pytools version 2021.2.9 unchanged (except for this message).
+https://github.com/inducer/pytools/blob/a0d9fdc33f459802b54a79c00a617fbc410e9d4d/pytools/persistent_dict.py.
+See https://snakemake.readthedocs.io/en/stable/project_info/faq.html#i-want-to-pass-variables-between-rules-is-that-possible
```

Comment:
I haven't used Snakemake and I'm sure there is good reason for this, but why does the file have to be copied here when the example uses `from pytools.persistent_dict import PersistentDict`?

Reply:
Sorry, didn't see your reasoning in the PR comment. That makes sense, closing this conversation.

Reply:
No worries - for a bit of background here: workflow specific packages are a bit of a pain at the moment. For example, we recently ran into when we added the `epiweeks` package to this pipeline, and...
* users have to upgrade their dependencies. Pretty normal, but problematic as there's so many ways to install nextstrain tools.
* it's not clear what set of packages should be installed -- e.g. "augur" environment vs "ncov" environment. Again, a range of different approaches here.
* the same nextstrain docker image is used to run all our builds, so the package needs to be added there

I know Tom & John have been thinking about this a lot (buildpacks etc), but I don't think changes are imminent.